### PR TITLE
Advantage computation in the RefActor

### PR DESCRIPTION
**Pull request by Darktex**
Run with:

```
python scripts/runnable_recipe_ray_vllm_weight_sync.py --config ../recipes/configs/dev/qwen3B_full_grpo.yaml
```

Things look good using the "default" arguments: 

```
# Fine-tuning arguments
batch_size: 16 # B*G
grpo_samples: 16
forward_batch_size: 1
max_generated_tokens: 512

# number of train_steps after which to do weight sync
steps_before_sync: 2
num_ref_workers: 1
num_fsdp_workers: 2

vllm:
  num_workers: 5
  tp_size: 1
  batch_size: 1 # TODO: for now, needs to match recipe batch size
  queue_maxsize: 3 #num workers * steps_before_sync
```
Screenshot 2025-03-31 at 5 59 07 PM
Screenshot 2025-03-31 at 5 59 33 PM

### CODE REVIEW

File: recipes/configs/dev/qwen3B_full_grpo.yaml [hunk 1]
```diff
@@ -71,8 +71,8 @@ resume_from_checkpoint: False
 save_every_n_epochs: 1
 
 # Fine-tuning arguments
-batch_size: 1
-grpo_samples: 16
+batch_size: 6
+grpo_samples: 5
```
Comment (felipemello1) [inline recipes/configs/dev/qwen3B_full_grpo.yaml:75]: 1. Shouldnt we have one for RefActor too?
2. super cool with leaving it like this, but at some point we might want to do what we did for vllm, and have it actor have its own section

Comment (Darktex) [reply recipes/configs/dev/qwen3B_full_grpo.yaml:75]: Great call. Makes sense to me, perhaps this can even be applied at the actor level (once we start factoring out common components?)

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 2]
```diff
@@ -274,39 +284,24 @@ def run(self):
                     if self._is_actor_zero:
                         print(f"Getting from rollout_queue queue.")
                     trajectory = self.rollout_queue.get(timeout=0.5)
+                    trajectory = trajectory.to(self._device)
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:287]: most likely, this is optimal, but we may not need for every content to be in the gpu, e.g metadata.

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 3]
```diff
@@ -274,39 +284,24 @@ def run(self):
...
             masks = generation.get_causal_mask_from_padding_mask(
-                query_response_padding_masks
+                trajectory["query_response_padding_masks"]
             )
             position_ids = generation.get_position_ids_from_padding_mask(
-                query_response_padding_masks
+                trajectory["query_response_padding_masks"]
             )
 
             # Reset GPU memory stats before model_running
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:307]: i think we can delete this. Our utility already resets it.

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 4]
```diff
@@ -315,33 +310,89 @@ def run(self):
             time_grpo_steps_start = time.perf_counter()
             with torch.no_grad():
                 ref_logits = self._ref_model(
-                    query_responses, input_pos=position_ids, mask=masks
+                    trajectory["query_responses"], input_pos=position_ids, mask=masks
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:313]: did you decide to keep it named as "query_responses"? should we change to something more obvious?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 5]
```diff
@@ -315,33 +310,89 @@ def run(self):
...
 
             del ref_logits, position_ids, masks
             # masking of ref_logprobs is done in grpo_step
 
+            # Extract components from raw trajectory
+            query_responses = trajectory["query_responses"]
+            responses = trajectory["responses"]
+            query_response_padding_masks = trajectory["query_response_padding_masks"]
+            seq_lens = trajectory["seq_lens"]
+            answers = trajectory["answers"]
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:331]: if we are mostly doing trajectory["key"] in the code before it, should we keep the same pattern and avoid unwrapping the trajectory?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 6]
```diff
@@ -315,33 +310,89 @@ def run(self):
...
+            padded_tokens_percentage = (
+                (padded_tokens / total_tokens) * 100 if total_tokens > 0 else 0
+            )
+            number_of_tokens = seq_lens.sum().item()
+
+            # Truncate sequences at first stop token
+            response_padding_masks, responses = (
+                rlhf.truncate_sequence_at_first_stop_token(
+                    responses,
+                    torch.tensor(self._tokenizer.stop_tokens, device=self._device),
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:345]: creating this tensor every time seems wasteful. Maybe we can move it elsewhere?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 7]
```diff
@@ -315,33 +310,89 @@ def run(self):
...
+
+            # Compute rewards
+            responses = responses.reshape(
+                batch_size // self.grpo_samples, self.grpo_samples, -1
+            )
+            rewards_full, successes_full, reward_metadata = batched_rewards(
+                self._tokenizer, responses, answers, device=self._device
+            )
+
+            # Compute advantages: B, G, num_funcs -> B, G
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:368]: there was an all_reduce function here before, for the sake of logging the sum of rewards across all actors. Should we migrate the logging of rewards/successes to RefActor?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 8]
```diff
@@ -608,14 +659,14 @@ def postprocess_vllm_request_output(request_output):
...
-                seq_lens,
-            ]
+            return TensorDict(
+                query_responses=query_responses,
+                responses=responses,
+                logprobs=logprobs,
+                query_response_padding_masks=query_response_padding_masks,
+                seq_lens=seq_lens,
+                batch_size=bs,
+            )
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:669]: awesome

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 9]
```diff
@@ -1324,11 +1359,15 @@ def train(self):
...
+            raw_sample = self.replay_buffer.sample(batch_size=self.batch_size)
+            print(f"PyTorchActor samples: {raw_sample}")
+
             trajectory = self.replay_buffer.sample(batch_size=self.batch_size)[0].to(
                 self._device
             )
             time_waiting_buffer = time.perf_counter() - time_waiting_buffer_start
-            log.info(f"{self.rank=} got from queue traj {trajectory}")
+            if self._is_rank_zero:
+                log.info(f"{self.rank=} got from queue traj {trajectory}")
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:1370]: we have utils.log_rank_zero, so you dont have to do `if self._is_rank_zero:`

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 10]
```diff
@@ -1529,24 +1559,10 @@ def _prepare_trajectory(self, raw_trajectory):
...
-                successes_mean
-            ) = None
-
-        # Clean up intermediate variables
-        del rewards_full, successes_full, responses, rewards_sum
-        torch.distributed.barrier()
+        rewards_mean_per_func = rewards.mean(dim=0).cpu()
+        successes_mean_per_func = successes.mean(dim=0).cpu()
+        rewards_mean = rewards_mean_per_func.mean()
+        successes_mean = successes_mean_per_func.mean()
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:1565]: this means that the reward we log will be only from rank 0. I am ok with that. But does it make sense to you?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 11]
```diff
@@ -315,33 +310,89 @@ def run(self):
             time_grpo_steps_start = time.perf_counter()
             with torch.no_grad():
                 ref_logits = self._ref_model(
-                    query_responses, input_pos=position_ids, mask=masks
+                    trajectory["query_responses"], input_pos=position_ids, mask=masks
                 )
             time_model_running = time.perf_counter() - time_grpo_steps_start
 
             ref_logits = rlhf.truncate_sequence_for_logprobs(ref_logits, context_length)
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:317]: This assumes batched prompts have been padded to a fixed length so `context_length` is fixed for every sample
```python
def truncate_sequence_for_logprobs(
    query_response_logits: torch.Tensor, context_length: int
) -> torch.Tensor:
    """
    Truncates logits generated over a sequence for estimating logprobs over the tokens in the sequence.
    This assumes the sequence is of the (query, response) format with length (context_length + response_length)
    Args:
        query_response_logits (torch.Tensor): The logits tensor of shape [b, context_length + response_length, vocab_size].
        context_length (int): The length of the context.

    Returns:
        torch.Tensor: The truncated logits for the response with shape [b, response_length, vocab_size]."""
    return query_response_logits[:, context_length - 1 : -1]
```

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 12]
```diff
@@ -315,33 +310,89 @@ def run(self):
...
+            responses = responses.reshape(
+                batch_size // self.grpo_samples, self.grpo_samples, -1
+            )
+            rewards_full, successes_full, reward_metadata = batched_rewards(
+                self._tokenizer, responses, answers, device=self._device
+            )
+
+            # Compute advantages: B, G, num_funcs -> B, G
+            rewards_sum = rewards_full.sum(-1)
+            advantages = (rewards_sum - rewards_sum.mean(1, keepdim=True)) / (
```
Comment (SalmanMohammadi) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:370]: Have the rewards already been gathered by this point?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 4 above)
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:313]: Let's do this as part of a refactor

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 5 above)
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:331]: Yeah but we need to refactor this whole portion anyway :D

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 12 above)
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:370]: There is no torch.distributed group in the RefActor (well, at least not for now), so no need to do that

**Comment by Darktex** [comment]
Since last time, made a few changes:

1. Moved the advantage and reward logging into the RefActor accordingly
2. Added more explicit shapes
3. Changed the reward computation function to take answers of size (B, G) and manually duplicated answers. This is a first step to being super deliberate about shapes, with more refactoring to come.

Issues: 

I don't know if it's the logging or what, but I'm not getting good curves from my runs. 

I could use help in checking what's going on here. I followed the same settings as Felipe:

```
# Fine-tuning arguments
batch_size: 1
grpo_samples: 16
forward_batch_size: 1
max_generated_tokens: 512

# number of train_steps after which to do weight sync
steps_before_sync: 2
num_ref_workers: 1
num_fsdp_workers: 2

vllm:
  num_workers: 5
  tp_size: 1
  batch_size: 1 # TODO: for now, needs to match recipe batch size
  queue_maxsize: 3 #num workers * steps_before_sync
```

Screenshot 2025-03-31 at 5 34 23 PM

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 2 above)
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:319]: GPUs can't hold strings anyway, so that should be fine

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:349]: It's already the case, right? We'll do packing later

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 7 above)
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:405]: Logging migrated

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 8 above)
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:720]: A lot more to do...

File: recipes/configs/dev/qwen3B_full_grpo.yaml [hunk 13]
```diff
@@ -116,7 +115,7 @@ num_fsdp_workers: 2
 vllm:
   num_workers: 5
   tp_size: 1
-  batch_size: 1 # TODO: for now, needs to match recipe batch size
+  batch_size: 1
```
Comment (Darktex) [inline recipes/configs/dev/qwen3B_full_grpo.yaml:118]: Keeping these values like this for now, but we need to move away from using 1 as a number... like ANYWHERE :D

**Review by vmoens — APPROVED** [review]
LGTM!

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 14]
```diff
@@ -274,39 +320,24 @@ def run(self):
...
-                responses,
-                logprobs,
-                query_response_padding_masks,
-                seq_lens,
-                answers,
-                policy_version,
-            ) = trajectory
-
-            context_length = query_responses.shape[1] - responses.shape[1]
+            context_length = (
```
Comment (SalmanMohammadi) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:331]: When will this not just be `context_length = responses.shape[1]`?

**Review by SalmanMohammadi — APPROVED** [review]
